framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,372 @@
+    // For loading from a Jar file
+    private static final String CLASS_SUFFIX = ".class";
+
+    // For loading from external directories
+    private static final String JAVA_SUFFIX = ".java";
+
+    // Constants
+    private static final char DOT = '.';
+    private static final char SLASH = '/';
+
+    /**
+     * Processing Env used to create an Annotation Builder, which is in turn
+     * used to build the annotation mirror from the loaded Class object
+     */
+    protected ProcessingEnvironment processingEnv;
+
+    // Stores the Resource URL of the qual directory of a Checker class, or AnnotatedTypeFactory class of a particular checker
+    URL resourceURL;
+
+    // Stores a mapping of the Annotation object and the converted AnnotationMirror
+    private Map<Class<? extends Annotation>, AnnotationMirror> loadedAnnotations;
+
+    /** 
+     * constructor for loading type qualifiers defined for a checker
+     * @param checker a base type checker or its subclass
+     */
+    public AnnotatedTypeLoader(BaseTypeChecker checker) {
+        this.checker = checker;
+        processingEnv = checker.getProcessingEnvironment();
+        packageName = checker.getClass().getPackage().getName().replace(SLASH, DOT) + QUAL_PACKAGE_SUFFIX;
+        ProcessPackageNameAndLoadAnnos();
+    }
+
+    /**
+     * Constructor for loading type qualifiers defined for an annotated type factory
+     * @param factory an annotated type factory or its subclass
+     */
+    public AnnotatedTypeLoader(AnnotatedTypeFactory factory) {
+        processingEnv = factory.getProcessingEnv();
+        packageName = factory.getClass().getPackage().getName().replace(SLASH, DOT) + QUAL_PACKAGE_SUFFIX;
+        ProcessPackageNameAndLoadAnnos();
+    }
+
+    // sees if the package name of the qual directory is a valid resource and
+    // retrieve its resource URL (either a jar or regular file directory), then
+    // proceeds to reflectively load annotations if it is valid. Returns an
+    // empty set if it isn't a valid URL
+    private void ProcessPackageNameAndLoadAnnos() {
+        //resourceURL = Thread.currentThread().getContextClassLoader().getResource(packageName.replace(DOT, SLASH));
+        resourceURL = this.getClass().getClassLoader().getResource(packageName.replace(DOT, SLASH));
+
+        // in checkers, there will be a resource URL for the qual directory. But
+        // when called in the framework (eg GeneralAnnotatedTypeFactory), there
+        // won't be a resourceURL since there isn't a qual directory.
+        if (resourceURL != null) {
+            // if there's a URL, load from it
+            loadedAnnotations = loadCheckerFrameworkAnnotations();
+        } else {
+            // otherwise create an empty set
+            loadedAnnotations = new HashMap<Class<? extends Annotation>, AnnotationMirror>();
+        }
+    }
+
+    /**
+     * Every subclass of AnnotatedTypeLoader must implement how it will convert a class into an annotation mirror
```

Comment:
I think "must" here is a little strong.   The comment should also include say something about how this method can be used to filter annotations.